File: code/R/chapter05/lmvnpdf.R
# lmvnpdf.R
# Evaluates the log density of a multivariate normal
# distribution
#
# Inputs:
# x: evaluation points, k x 1
# mu: mean vector, k x 1
# Sig: covariance matrix, k x k
#
# Output:
# logden: log density of N(mu,Sig) evaluated at x

lmvnpdf <- function(x, mu, Sig) {
    x <- as.numeric(x)
    mu <- as.numeric(mu)
    k <- length(mu)

    # R's chol() returns the UPPER factor, so transpose it to obtain
    # the lower Cholesky factor of MATLAB's chol(Sig,'lower')
    CSig <- t(chol(as.matrix(Sig)))
    e <- forwardsolve(CSig, x - mu)

    logden <- -0.5*k*log(2*pi) - sum(log(diag(CSig))) - 0.5*sum(e^2)
    logden
}

File: code/python/chapter05/lmvnpdf.py
# lmvnpdf.py
# Evaluates the log density of a multivariate normal distribution.

import numpy as np
from scipy.linalg import solve_triangular


def lmvnpdf(x, mu, Sig):
    """Log density of N(mu, Sig) evaluated at x.

    Inputs:
      x  : evaluation point, k-vector
      mu : mean vector, k-vector
      Sig: covariance matrix, k x k

    Output:
      logden: log density of N(mu, Sig) evaluated at x
    """
    x = np.asarray(x).flatten()
    mu = np.asarray(mu).flatten()
    k = len(mu)

    CSig = np.linalg.cholesky(Sig)  # lower Cholesky factor
    e = solve_triangular(CSig, x - mu, lower=True)

    logden = -0.5*k*np.log(2*np.pi) - np.sum(np.log(np.diag(CSig))) \
        - 0.5*(e @ e)
    return logden
